File: src/main/java/fi/metatavu/oioi/cm/rest/AbstractApi.java
package fi.metatavu.oioi.cm.rest;

import java.security.Principal;
import java.util.Collection;
import java.util.Collections;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.core.Response;

import org.jboss.resteasy.spi.ResteasyProviderFactory;
import org.keycloak.KeycloakPrincipal;
import org.keycloak.KeycloakSecurityContext;
import org.keycloak.authorization.client.AuthzClient;
import org.keycloak.authorization.client.ClientAuthorizationContext;
import org.keycloak.representations.AccessToken;
import org.keycloak.representations.AccessToken.Access;
import org.slf4j.Logger;

import fi.metatavu.oioi.cm.model.ErrorResponse;

/**
 * Abstract base class for all API services
 * 
 * @author Antti Leppä
 * @author Heikki Kurhinen
 */
public abstract class AbstractApi {
  
  protected static final String NOT_FOUND_MESSAGE = "Not found";
  protected static final String CUSTOMER_DEVICE_MISMATCH_MESSAGE = "Device does not belong to this customer";
  protected static final String APPLICATION_DEVICE_MISMATCH_MESSAGE = "Application does not belong to this device";
  protected static final String FORBIDDEN_MESSAGE = "Forbidden";

  protected static final String ADMIN_ROLE = "admin";
  
  @Inject
  private Logger logger;
  
  /**
   * Constructs ok response
   * 
   * @param entity payload
   * @return response
   */
  protected Response createOk(Object entity) {
    return Response
      .status(Response.Status.OK)
      .entity(entity)
      .build();
  }

  /**
   * Constructs ok or not found response if entity is null
   * 
   * @param entity payload
   * @return response
   */
  protected Response createOkOrNotFound(Object entity) {
    if (entity == null) {
      return createNotFound(NOT_FOUND_MESSAGE);
    }

    return createOk(entity);
  }
  
  /**
   * Constructs not found response
   * 
   * @param message message
   * @return response
   */
  protected Response createNotFound(String message) {
    ErrorResponse entity = new ErrorResponse();
    entity.setMessage(message);
    return Response
      .status(Response.Status.NOT_FOUND)
      .entity(entity)
      .build();
  }
  
  /**
   * Constructs forbidden response
   * 
   * @param message message
   * @return response
   */
  protected Response createForbidden(String message) {
    ErrorResponse entity = new ErrorResponse();
    entity.setMessage(message);
    return Response
      .status(Response.Status.FORBIDDEN)
      .entity(entity)
      .build();
  }
  
  /**
   * Constructs bad request response
   * 
   * @param message message
   * @return response
   */
  protected Response createBadRequest(String message) {
    ErrorResponse entity = new ErrorResponse();
    entity.setMessage(message);
    return Response
      .status(Response.Status.BAD_REQUEST)
      .entity(entity)
      .build();
  }
  
  /**
   * Constructs internal server error response
   * 
   * @param message message
   * @return response
   */
  protected Response createInternalServerError(String message) {
    ErrorResponse entity = new ErrorResponse();
    entity.setMessage(message);
    return Response
      .status(Response.Status.INTERNAL_SERVER_ERROR)
      .entity(entity)
      .build();
  }
  
  
  /**
   * Constructs ok response
   * 
   * @param entity payload
   * @param totalHits total hits
   * @return response
   */
  protected Response createOk(Object entity, Long totalHits) {
    return Response
      .status(Response.Status.OK)
      .entity(entity)
      .header("Total-Results", totalHits)
      .build();
  }
  
  /**
   * Constructs no content response
   * 
   * @return response
   */
  protected Response createNoContent() {
    return Response
      .status(Response.Status.NO_CONTENT)
      .build();
  }
  
  /**
   * Returns logged user id
   * 
   * @return logged user id
   */
  protected UUID getLoggerUserId() {
    HttpServletRequest httpServletRequest = getHttpServletRequest();
    String remoteUser = httpServletRequest.getRemoteUser();
    if (remoteUser == null) {
      return null;
    }
    
    return UUID.fromString(remoteUser);
  }
  
  /**
   * Returns request locale
   * 
   * @return request locale
   */
  protected Locale getLocale() {
    return getHttpServletRequest().getLocale();
  }
  
  /**
   * Return current HttpServletRequest
   * 
   * @return current http servlet request
   */
  protected HttpServletRequest getHttpServletRequest() {
    return ResteasyProviderFactory.getContextData(HttpServletRequest.class);
  }

  /**
   * Return Keycloak authorization client
   */
  protected AuthzClient getAuthzClient() {
    ClientAuthorizationContext clientAuthorizationContext = getAuthorizationContext();
    if (clientAuthorizationContext == null) {
      logger.error("Failed to retrieve Keycloak authorization client");
      return null;
    }

    return clientAuthorizationContext.getClient();
  }

  /**
   * Returns whether logged user has role
   * 
   * @param role role
   * @return whether logged user has specified role or not
   */
  protected Boolean hasRealmRole(String roleName) {
    KeycloakSecurityContext keycloakSecurityContext = getKeycloakSecurityContext();
    if (keycloakSecurityContext == null) {
      System.out.println("Cannot get security context");
      return false;
    }

    AccessToken token = keycloakSecurityContext.getToken();
    if (token == null) {
      System.out.println("Cannot get token");
      return false;
    }

    Access realmAccess = token.getRealmAccess();
    if (realmAccess == null) {
      System.out.println("Cannot get realm access");
      return false;
    }

    System.out.println("User roles: ");
    realmAccess.getRoles().stream().forEach(r -> System.out.println(r));
    return realmAccess.isUserInRole(roleName);
  }
  
  /**
   * Returns group ids where logged user is a member of
   * 
   * @return group ids where logged user is a member of
   */
  protected Set<String> getLoggedUserGroups() {
    KeycloakSecurityContext context = getKeycloakSecurityContext();
    if (context == null) {
      return Collections.emptySet();
    }

    AccessToken token = context.getToken();
    if (token == null) {
      return Collections.emptySet();
    }
    
    Map<String, Object> otherClaims = context.getToken() .getOtherClaims();
    if (otherClaims == null) {
      return Collections.emptySet();
    }
    
    @SuppressWarnings("unchecked") 
    Collection<String> groupIds = (Collection<String>) otherClaims.get("groups");
    if (groupIds == null) {
      return Collections.emptySet();
    }
    
    return groupIds.stream().map(group -> group).collect(Collectors.toSet());    
  }

  protected boolean isAdminOrHasCustomerGroup(String customerName) {
    Boolean isAdmin = hasRealmRole(ADMIN_ROLE);
    if (isAdmin) {
      return true;
    }

    Set<String> groups = getLoggedUserGroups();
    return groups.contains(customerName);
  }

  /**
   * Return keycloak authorization client context or null if not available 
   */
  private ClientAuthorizationContext getAuthorizationContext() {
    KeycloakSecurityContext keycloakSecurityContext = getKeycloakSecurityContext();
    if (keycloakSecurityContext == null) {
      logger.error("Failed to retrieve KeycloakSecurityContext");
      return null;
    }

    return (ClientAuthorizationContext) keycloakSecurityContext.getAuthorizationContext();
  }

  /**
   * Returns keycloak security context from request or null if not available
   */
  private KeycloakSecurityContext getKeycloakSecurityContext() {
    HttpServletRequest request = getHttpServletRequest();
    Principal userPrincipal = request.getUserPrincipal();
    KeycloakPrincipal<?> kcPrincipal = (KeycloakPrincipal<?>) userPrincipal;
    if (kcPrincipal == null) {
      return null;
    }
    
    return kcPrincipal.getKeycloakSecurityContext();
  }
}

File: src/main/kotlin/fi/metatavu/oioi/cm/rest/AbstractApi.kt
package  fi.metatavu.oioi.cm.rest

import fi.metatavu.oioi.cm.model.ErrorResponse
import io.quarkus.security.identity.SecurityIdentity
import org.eclipse.microprofile.config.inject.ConfigProperty
import org.eclipse.microprofile.jwt.JsonWebToken
import org.keycloak.authorization.client.AuthzClient
import org.keycloak.authorization.client.Configuration
import org.slf4j.Logger
import java.util.*
import javax.enterprise.context.RequestScoped
import javax.inject.Inject
import javax.ws.rs.core.Context
import javax.ws.rs.core.HttpHeaders
import javax.ws.rs.core.Response

/**
 * Abstract base class for all API services
 *
 * @author Antti Leppä
 */
@RequestScoped
abstract class AbstractApi {

    @Inject
    lateinit var logger: Logger

    @Inject
    lateinit var jsonWebToken: JsonWebToken

    @Inject
    lateinit var securityIdentity: SecurityIdentity

    @ConfigProperty(name = "oioi.keycloak.url")
    lateinit var keycloakUrl: String

    @ConfigProperty(name = "oioi.keycloak.realm")
    lateinit var keycloakRealm: String

    @ConfigProperty(name = "quarkus.oidc.client-id")
    lateinit var keycloakClientId: String

    @ConfigProperty(name = "quarkus.oidc.credentials.secret")
    lateinit var keycloakSecret: String

    @Context
    lateinit var headers: HttpHeaders

    /**
     * Returns used api key or null if not available
     */
    protected val apiKey: String?
        get() {
            return headers.getHeaderString("X-API-KEY")
        }

    /**
     * Returns logged user id
     *
     * @return logged user id
     */
    protected val loggedUserId: UUID?
        get() {
            if (jsonWebToken.subject != null) {
                return try {
                    UUID.fromString(jsonWebToken.subject)
                } catch (ex: IllegalArgumentException) {
                    logger.error(ex.message)
                    null
                }
            }

            return null
        }

    /**
     * Returns whether user has given realm role
     *
     * @param role role
     * @return whether user has given realm role
     */
    protected fun hasRealmRole(role: String): Boolean {
        return securityIdentity.hasRole(role)
    }

    /**
     * Returns whether logged user is either admin or customer admin
     *
     * @param customerName customer's name
     * @return whether logged user is either admin or customer admin
     */
    protected fun isCustomerAdmin(customerName: String?): Boolean {
        if (hasRealmRole(ADMIN_ROLE)) {
            return true
        }

        customerName ?: return false
        val roleName = "$customerName-admin"
        return hasRealmRole(roleName) || loggedUserGroups.contains(roleName)
    }

    /**
     * Returns whether logged user is either admin or has customer group membership
     *
     * @param customerName customer's name
     * @return whether logged user is either admin or has customer group membership
     */
    protected fun isAdminOrHasCustomerGroup(customerName: String?): Boolean {
        if (isCustomerAdmin(customerName = customerName)) {
            return true
        }

        customerName ?: return false

        return hasRealmRole(customerName) || loggedUserGroups.contains(customerName)
    }

    /**
     * Returns logged user's user groups
     *
     * @returns logged user's user groups
     */
    protected val loggedUserGroups: Set<String>
        get() {
            return jsonWebToken.groups ?: emptySet()
        }

    /**
     * Return keycloak authorization client
     */
    protected val authzClient: AuthzClient
        get() {
           val configuration = Configuration()
            configuration.realm = keycloakRealm
            configuration.isBearerOnly = true
            configuration.authServerUrl = keycloakUrl
            configuration.sslRequired = "external"
            configuration.resource = keycloakClientId
            configuration.confidentialPort = 0
            configuration.credentials["secret"] = keycloakSecret
            return AuthzClient.create(configuration)
        }

    /**
     * Constructs ok response
     *
     * @param entity payload
     * @return response
     */
    protected fun createOk(entity: Any?): Response {
        return Response
            .status(Response.Status.OK)
            .entity(entity)
            .build()
    }

    /**
     * Constructs no content response
     *
     * @param entity payload
     * @return response
     */
    protected fun createAccepted(entity: Any?): Response {
        return Response
            .status(Response.Status.ACCEPTED)
            .entity(entity)
            .build()
    }

    /**
     * Constructs no content response
     *
     * @return response
     */
    protected fun createNoContent(): Response {
        return Response
            .status(Response.Status.NO_CONTENT)
            .build()
    }

    /**
     * Constructs bad request response
     *
     * @param message message
     * @return response
     */
    protected fun createBadRequest(message: String): Response {
        return createError(Response.Status.BAD_REQUEST, message)
    }

    /**
     * Constructs not found response
     *
     * @param message message
     * @return response
     */
    protected fun createNotFound(message: String): Response {
        return createError(Response.Status.NOT_FOUND, message)
    }

    /**
     * Constructs not found response
     *
     * @return response
     */
    protected fun createNotFound(): Response {
        return Response
            .status(Response.Status.NOT_FOUND)
            .build()
    }


    /**
     * Constructs not found response
     *
     * @param message message
     * @return response
     */
    protected fun createConflict(message: String): Response {
        return createError(Response.Status.CONFLICT, message)
    }

    /**
     * Constructs not implemented response
     *
     * @param message message
     * @return response
     */
    protected fun createNotImplemented(message: String): Response {
        return createError(Response.Status.NOT_IMPLEMENTED, message)
    }

    /**
     * Constructs internal server error response
     *
     * @param message message
     * @return response
     */
    protected fun createInternalServerError(message: String): Response {
        return createError(Response.Status.INTERNAL_SERVER_ERROR, message)
    }

    /**
     * Constructs forbidden response
     *
     * @param message message
     * @return response
     */
    protected fun createForbidden(message: String): Response {
        return createError(Response.Status.FORBIDDEN, message)
    }

    /**
     * Constructs unauthorized response
     *
     * @param message message
     * @return response
     */
    protected fun createUnauthorized(message: String): Response {
        return createError(Response.Status.UNAUTHORIZED, message)
    }

    /**
     * Constructs an error response
     *
     * @param status status code
     * @param message message
     *
     * @return error response
     */
    private fun createError(status: Response.Status, message: String): Response {
        val entity = ErrorResponse(
            message = message
        )

        return Response
            .status(status)
            .entity(entity)
            .build()
    }

    companion object {
        const val NOT_FOUND_MESSAGE = "Not found"
        const val CUSTOMER_DEVICE_MISMATCH_MESSAGE = "Device does not belong to this customer"
        const val APPLICATION_DEVICE_MISMATCH_MESSAGE = "Application does not belong to this device"
        const val UNAUTHORIZED = "Unauthorized"
        const val FORBIDDEN_MESSAGE = "Forbidden"
        const val ADMIN_ROLE = "admin"
        const val MISSING_PAYLOAD = "Missing payload"
    }

}